packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -1888,13 +1888,17 @@ abstract class MultiSelectableSelectionContainerDelegate extends SelectionContai
   /// [SelectWordSelectionEvent.globalPosition].
   @protected
   SelectionResult handleSelectWord(SelectWordSelectionEvent event) {
+    SelectionResult? lastSelectionResult;
     for (int index = 0; index < selectables.length; index += 1) {
       final Rect localRect = Rect.fromLTWH(0, 0, selectables[index].size.width, selectables[index].size.height);
       final Matrix4 transform = selectables[index].getTransformTo(null);
       final Rect globalRect = MatrixUtils.transformRect(transform, localRect);
       if (globalRect.contains(event.globalPosition)) {
         final SelectionGeometry existingGeometry = selectables[index].value;
-        dispatchSelectionEventToChild(selectables[index], event);
+        lastSelectionResult = dispatchSelectionEventToChild(selectables[index], event);
+        if (lastSelectionResult == SelectionResult.previous || lastSelectionResult == SelectionResult.next) {
```

Comment:
This seems weird, it should stop as soon as dispatchSelectionEventToChild does not return next.

imagine this case

in sequential order

selectable1 -> next (continue)
selectable2 -> next (continue)
selectable3 -> prev  (should stop)

this meant the selectword land in between selectable 2 and 3, one of them should be set to the currentSelectionStartIndex and currentSelectionEndIndex, I don't think it matter which one.

Reply:
That makes sense, but is this assuming that the selectables are in order of how the text is visually laid out?

Trying something like the following code, and the first selectable in the list is `Hello world this is some long text that goes\non two lines`. The second selectable is `Word`, and the third selectable is `Word Word2`.

If I change it to stop as soon as `dispatchSelectionEventToChild` does not return `next` then I get the following.

In sequential order this happens.

selectable1 -> prev (should stop)
currentSelectionStartIndex/currentSelectionEndIndex : null
should return SelectionResult.prev

Is this expected? Trying to right click `Word` at the beginning of the text fails since we never reach it.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(
        body: Center(
          child: SelectionArea(
            child: Text.rich(
              TextSpan(
                children: <InlineSpan>[
                  TextSpan(text: 'Word'),
                  WidgetSpan(child: SizedBox.shrink()),
                  TextSpan(text: ' Hello world this is some long text that goes\non two lines.'),
                  WidgetSpan(child: SizedBox.shrink()),
                  TextSpan(text: 'Word'),
                  TextSpan(text: ' Word2'),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

Reply:
> the selectables are in order of how the text is visually laid out?

they should be. I think the issue you encounter was a bug that the function to calculate screen order does not do well when the rect overlaps.
https://github.com/flutter/flutter/blob/764f4cc8c273b82316a221694ff522effd9d1138/packages/flutter/lib/src/widgets/selectable_region.dart#L1635